docs/rfcs/021-metering.md

```diff
@@ -0,0 +1,175 @@
+
+* Real storage size. That is the size of the tenant directory on the pageservers disk. Per-tenant.
+
+* S3 storage size. That is the size of the tenant data on S3. Per-tenant.
+
+That info should be enough to build an internal model that predicts AWS price (hence tracking `written data` and `real storage size`). As for the billing model we probably can get away with mentioning only `CPU time`, `synthetic storage size`, and `traffic` consumption.
+
+# Services participating in metrics collection
+
+## Proxy
+
+Proxy is the only place that knows about traffic flow, so it tracks it and reports it with quite a small interval, let's say 1 minute. A small interval is needed here since the proxy is stateless, and any restart will reset accumulated consumption. Also proxy should report deltas since the last report, not an absolute value of the counter. Such kind of events is easier to integrate over a period of time to get the amount of traffic during some time interval.
+
+Example event:
+
+```json
+{
+    "metric": "traffic_in_bytes",
+    "type": "increment",
+    "endpoint_id": "blazing-warrior-34",
+    "event_start_time": ...,
+    "event_stop_time": ...,
+    "value": 12345454,
+}
+```
+
+Since we report deltas over some period of time, it makes sense to include `event_start_time`/`event_stop_time` where `event_start_time` is the time of the previous report. That will allow us to identify metering gaps better (e.g., failed send/delivery).
+
+When there is no active connection proxy can avoid reporting anything. Also, deltas are additive, so several console instances serving the same user and endpoint can report traffic without coordination.
+
+## Console
+
+The console knows about start/stop events, so it knows the amount of CPU time allocated to each endpoint. It also knows about operation successes and failures and can avoid billing clients after unsuccessful 'suspend' events. The console doesn't know the current compute size within the allowed limits on the endpoint. So with CPU time, we do the following:
+
+* While we don't yet have the autoscaling console can report `cpu time` as the number of seconds since the last `start_compute` event.
+
+* When we have autoscaling, `autoscaler-agent` can report `cpu time`*`compute_units_count` in the same increments as the proxy reports traffic. 
+
+Example event:
+
+```json
+{
+    "metric": "effective_compute_seconds",
+    "type": "increment",
+    "endpoint_id": "blazing-warrior-34",
+    "event_start_time": ...,
+    "event_stop_time": ...,
+    "value": 12345454,
+}
+```
+
+I'd also suggest reporting one value, `cpu time`*`compute_units_count`, instead of two separate fields as it makes event schema simpler (it is possible to treat it the same way as traffic) and preserves additivity.
+
+## Pageserver
+
+Pageserver knows / has access to / can calculate the rest of the metrics:
```

Comment:
When should PageServer call control-plane about the tenant size? Can you provide kind of sequence diagram that shows the whole path from initiating the action to the final processing?

Reply:
More of less by using a timer. Period ideally is minutes, but some calculations there are expensive so for "synthetic" size it would be a bigger period (like 1 hour). We could later think on how to reduce that period.

Talking about the exact implementation details, most of the metrics described here could defined as Prometheus counters (except "synthetic" size) and updated on the fly. Then background job just collects all that info for all of the tenant and sends it to the console. Straightforward optimization here is not to send values that were not changed since last transmission.

Reply:
Taking counter value is not an expensive operation and we collect them each 15sec for all tenants by prometheus. Synthetic size should be taken more carefully though.

> PageServer will start the calculation with a timer for all tenants?

I didn't say that. Timer can trigger walk across all tenants calculating sizes one after another. Or do that with some amount of parallelism.